File: backend/app/api/splitting.py
from fastapi import APIRouter

from app.schemas.splitting import SplittingRequest, SplittingResponse
from app.services.splitting_service import split_dataset

router = APIRouter()


@router.post("/splitting/split", response_model=SplittingResponse)
def split_data(payload: SplittingRequest):
    # Endpoint to split dataset into training and testing sets.
    train_set, test_set, summary = split_dataset(
        headers=payload.dataset.headers,
        rows=payload.dataset.rows,
        test_size=payload.test_size,
        random_state=payload.random_state,
    )

    return SplittingResponse(
        train_set=train_set,
        test_set=test_set,
        summary=summary,
    )

File: frontend/services/splitting.ts
import { ParsedDataset } from "@/types/import";
import { SplittingResponse } from "@/types/splitting";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export async function fetchSplittingPreview(
  dataset: ParsedDataset,
  testSize: number = 0.2,
  randomState?: number,
): Promise<SplittingResponse> {
  const response = await fetch(`${API_BASE_URL}/api/splitting/split`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      dataset: {
        headers: dataset.headers,
        rows: dataset.rows,
      },
      test_size: testSize,
      random_state: randomState,
    }),
  });

  if (!response.ok) {
    throw new Error(
      `Failed to fetch splitting preview: ${response.statusText}`,
    );
  }

  const data = await response.json();

  // Convert snake_case to camelCase for consistency with frontend types
  return {
    trainSet: data.train_set,
    testSet: data.test_set,
    summary: {
      totalRows: data.summary.total_rows,
      trainRows: data.summary.train_rows,
      testRows: data.summary.test_rows,
      trainPercentage: data.summary.train_percentage,
      testPercentage: data.summary.test_percentage,
    },
  };
}
